EvanaPerveenOfficial/ToDo
# Todo 2 implement authentication endpoints

- Ticket(s): https://welltravel.atlassian.net/browse/UM-360

## Description
Implement following task,
- Setup fastapi project environment
- setup docker environment
- connect PostgreSQL to the project
- setup loguru to get logging management
- built user model and task model using sqlalchemy
- wrote validation models using pydantic
-  built token based authentication system using jwt

> - JWT: Use JWT token generation and verification.
> - Password Hashing: Implement bcrypt for secure password hashing.
> - Endpoints: Develop RESTful endpoints for user registration, login, and token.
> - Authorization Middleware: Develop middleware to enforce role-based access control.
> - Endpoint Permissions: Implement logic to restrict access to endpoints based on user roles

- configure SMTP email system to send verification mail to the user to confirm registration 
- created user profile and add CRUD to it
- wrote CRUD operations for task management.
- paginate tasks 
- implemented searching and filtering on tasks
- implemented task completion 
- Implemented full project on ReactJS

## Types of changes
- [ ] Build/DevOps related changes

## Testing guidelines
Run `docker compose up` and here is server links,
- Api documentation link: http://0.0.0.0:8000/docs/
Run `npm start` to see the frontend

## Deployment notes
- N/A

## ScreenShots

![Screenshot from 2024-05-26 14-35-15](https://github.com/EvanaPerveenOfficial/ToDo/assets/162853436/61d94864-870b-4f63-b4c0-a629da2e98bb)

![Screenshot from 2024-05-26 14-36-52](https://github.com/EvanaPerveenOfficial/ToDo/assets/162853436/f61f9706-fd4b-4861-abdd-77352b218719)

![Screenshot from 2024-05-26 14-37-21](https://github.com/EvanaPerveenOfficial/ToDo/assets/162853436/6d10b785-3663-4dbc-9cf1-7d910148edd1)


![Screenshot from 2024-05-26 14-38-05](https://github.com/EvanaPerveenOfficial/ToDo/assets/162853436/8d986d87-e8ec-418b-969a-5ff4032f70ac)


![Screenshot from 2024-05-26 14-39-39](https://github.com/EvanaPerveenOfficial/ToDo/assets/162853436/f10bfe94-753e-46e7-ab88-6be40f11e89f)

![Screenshot from 2024-05-26 14-40-51](https://github.com/EvanaPerveenOfficial/ToDo/assets/162853436/1e52c8a3-2dcf-4009-9da5-bb48825a6584)

![Screenshot from 2024-05-26 14-41-27](https://github.com/EvanaPerveenOfficial/ToDo/assets/162853436/65e82d88-381c-4301-82c0-155e3acb1b5d)

![Screenshot from 2024-05-26 14-42-08](https://github.com/EvanaPerveenOfficial/ToDo/assets/162853436/a970d28d-c745-446a-b128-373dc313618a)

![Screenshot from 2024-05-26 14-42-55](https://github.com/EvanaPerveenOfficial/ToDo/assets/162853436/733ba59d-7038-4e34-b163-6a94941ce6f4)

![Screenshot from 2024-05-26 14-43-45](https://github.com/EvanaPerveenOfficial/ToDo/assets/162853436/94266429-e2e9-47c1-81bd-ee6642a303eb)

![Screenshot from 2024-05-26 14-44-20](https://github.com/EvanaPerveenOfficial/ToDo/assets/162853436/5f40e8ac-ab0f-494c-9958-a6c17a1485e4)

![Screenshot from 2024-05-26 14-45-38](https://github.com/EvanaPerveenOfficial/ToDo/assets/162853436/e3c157d6-dd97-4689-8038-c49505b4f913)

![Screenshot from 2024-05-26 14-46-28](https://github.com/EvanaPerveenOfficial/ToDo/assets/162853436/a77e5789-680f-40e9-9bdf-94341efdb8ee)

![Screenshot from 2024-05-26 14-47-05](https://github.com/EvanaPerveenOfficial/ToDo/assets/162853436/b15619dd-97ec-420c-9b6b-e0bb340bf499)

![Screenshot from 2024-05-26 14-47-32](https://github.com/EvanaPerveenOfficial/ToDo/assets/162853436/bb4416ab-b56d-4d65-86aa-19b821c5fe12)

![Screenshot from 2024-05-26 14-47-58](https://github.com/EvanaPerveenOfficial/ToDo/assets/162853436/0c277293-625e-47b9-94e7-bce468c55d6e)

![Screenshot from 2024-05-26 14-48-25](https://github.com/EvanaPerveenOfficial/ToDo/assets/162853436/3bd2b080-155c-47ff-b6d4-4e1a4805ef98)
